### PR TITLE
NIFI-3380 Support Multiple Versions of the Same Component

### PR DESCRIPTION
**Pull request by bbende**
This PR introduces support for running multiple versions of the same component. 

In order to make full use of this capability we need to release and updated version of the NAR Maven plugin (currently 1.1.0, need to release 1.2.0) and then update NiFi's master branch to use this new version. In this PR we have left it using the 1.1.0 NAR plugin, and NiFi is able to run as it normally does, but you can't actually run multiple versions of the same component yet because the framework needs the new information from the new NAR plugin.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/pom.xml [hunk 1]
```diff
@@ -440,6 +440,7 @@
                                                 <include>${staging.dir}/js/nf/canvas/nf-clipboard.js</include>
                                                 <include>${staging.dir}/js/nf/canvas/nf-custom-ui.js</include>
                                                 <include>${staging.dir}/js/nf/canvas/nf-canvas-utils.js</include>
+                                                <include>${staging.dir}/js/nf/canvas/nf-component-version.js</include>
```
Comment (scottyaslan) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/pom.xml]: I think we can move this down in the order. If you move it below the nf-process-group-configuration.js. then in the nfProcessGroupConfiguration module you can load nfComponentVersion module. If you do this then you can avoid having to inject the nfProcessGroupConfiguration module into the nfComponentVersion module in the nf-canvas-bootstrap.js.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-canvas-bootstrap.js [hunk 2]
```diff
@@ -339,6 +342,7 @@
                     nfActions.init();
                     nfQueueListing.init();
                     nfComponentState.init();
+                    nfComponentVersion.init(nfSettings, nfProcessGroupConfiguration);
```
Comment (scottyaslan) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-canvas-bootstrap.js]: Should be able to remove the injection of the nfProcessGroupConfiguration module here.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-component-version.js [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        // update the tags and description
+        $('#component-version-tags').text(documentedType.tags.join(', '));
+        $('#component-version-description').text(documentedType.description);
+    };
+
+    return {
+        init: function (settings, processGroupConfiguration) {
+            versionMap = d3.map();
+            nfSettings = settings;
+            nfProcessGroupConfiguration = processGroupConfiguration;
```
Comment (scottyaslan) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-component-version.js]: Should be able to remove the injection of the nfProcessGroupConfiguration module here too...

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 4]
```diff
@@ -1214,6 +1217,23 @@
...
+        changeVersion: function (selection) {
+            if (selection.size() !== 1 || !nf.CanvasUtils.isProcessor(selection)) {
+                return;
+            }
+
+            // get the processor data
+            var processor = selection.datum();
+
+            // attempt to change the version of the specified component
+            nf.ComponentVersion.promptForVersionChange(processor);
```
Comment (scottyaslan) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]: We should not be calling anything from the global scope. Instead we should have this module require('nf.ComponentVersion')...

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-component-version.js [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+ * Views state for a given component.
+ */
+(function (root, factory) {
+    if (typeof define === 'function' && define.amd) {
+        define(['jquery',
+                'nf.ErrorHandler',
+                'nf.Common',
+                'nf.Client',
+                'nf.CanvasUtils',
+                'nf.ng.Bridge'],
```
Comment (scottyaslan) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-component-version.js]: This module should require('nf.ProcessGroupConfiguration')...

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/pom.xml (hunk 1 above)
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/pom.xml]: Got it. I feel like it wasn't possible when I originally typed it up. However, this module went through a couple iterations during the development of this feature. I'll try updating the dependency injection with your suggestions now.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 4 above)
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]: Yup. Must have missed this one. Much of this was written prior to getting the modularization effort in. I went back through and updated each invocation using `nf.`. I'll update this one and make sure I didn't miss any others.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/resources/filters/canvas.properties [hunk 6]
```diff
@@ -28,6 +28,7 @@ nf.canvas.script.tags=<script type="text/javascript" src="js/nf/nf-ng-bridge.js?
 <script type="text/javascript" src="js/nf/canvas/nf-clipboard.js?${project.version}"></script>\n\
 <script type="text/javascript" src="js/nf/canvas/nf-custom-ui.js?${project.version}"></script>\n\
 <script type="text/javascript" src="js/nf/canvas/nf-canvas-utils.js?${project.version}"></script>\n\
+<script type="text/javascript" src="js/nf/canvas/nf-component-version.js?${project.version}"></script>\n\
```
Comment (scottyaslan) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/resources/filters/canvas.properties]: Will need to update the load order here as well...

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/jquery/propertytable/jquery.propertytable.js [hunk 7]
```diff
@@ -939,27 +939,30 @@
     /**
      * Gets the available controller services that implement the specified type and
      * prompts the user to create one.
```
Comment (scottyaslan) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/jquery/propertytable/jquery.propertytable.js]: Why did you delete the param docs?

Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/jquery/propertytable/jquery.propertytable.js]: Yeah, werid... Will restore.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/resources/filters/canvas.properties (hunk 6 above)
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/resources/filters/canvas.properties]: Yup.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardProcessorNode.java [hunk 8]
```diff
@@ -857,7 +858,13 @@ public Relationship getRelationship(final String relationshipName) {
 
     @Override
     public Processor getProcessor() {
-        return this.processor;
+        return processorRef.get().getProcessor();
+    }
+
+    @Override
+    public void setProcessor(final LoggableComponent<Processor> processor) {
+        final ProcessorDetails processorDetails = new ProcessorDetails(processor);
```
Comment (markap14) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardProcessorNode.java]: We should probably have some thread protection here to make sure that the Processor is indeed stopped. This would mean that we need to enter a synchronized block here, then verify that the processor is not running (!isRunning()) and then do the processorRef.set(processorDetails). We would also need to ensure that in the start() method that the act of changing the scheduledState from STOPPED to STARTED is also synchronized to ensure atomicity.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardProcessorNode.java [hunk 9]
```diff
@@ -1457,4 +1466,85 @@ public String getProcessGroupIdentifier() {
         final ProcessGroup group = getProcessGroup();
         return group == null ? null : group.getIdentifier();
     }
+
+    private static final class ProcessorDetails {
```
Comment (markap14) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardProcessorNode.java]: Would recommend that this be externalized into its own class file, since this class already is at nearly 1500 lines of code, and this inner class is static. Not a big deal but worth considering.

Comment (markap14) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardProcessorNode.java]: Similarly, for ControllerServiceDetails and ReportingTaskDetails

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/reporting/AbstractReportingTaskNode.java [hunk 10]
```diff
@@ -125,7 +141,12 @@ public void setSchedulingPeriod(final String schedulingPeriod) {
 
     @Override
     public ReportingTask getReportingTask() {
-        return reportingTask;
+        return reportingTaskRef.get().getReportingTask();
+    }
+
+    @Override
+    public void setReportingTask(final LoggableComponent<ReportingTask> reportingTask) {
+        this.reportingTaskRef.set(new ReportingTaskDetails(reportingTask));
```
Comment (markap14) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/reporting/AbstractReportingTaskNode.java]: We should make sure here that the Reporting Task is stopped, similar to Processors. And should also ensure that Controller Services are disabled when calling setControllerService.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardFlowService.java [hunk 11]
```diff
@@ -905,6 +932,14 @@ private void loadFromConnectionResponse(final ConnectionResponse response) throw
 
     }
 
+    private void initialzeController() throws IOException {
```
Comment (markap14) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardFlowService.java]: Typo in method name - initialze should be initialize

**Comment by markap14** [comment]
@bbende I created a 3-node cluster. I added MyProcessor v2.0 to the flow. I then shut down node 3, removed the nifi-example-processors-nar-2.0.nar file from the lib/ directory and removed conf/flow.xml.gz. I then restarted Node 3. It fails to startup because the cluster flow cannot be inherited, due to the node not having v2.0 of the Processor. This, I think, is the expected behavior. However, it fails with the following error message in the logs:

```
2017-03-17 13:14:58,602 ERROR [main] o.a.nifi.controller.StandardFlowService Failed to load flow from cluster due to: org.apache.nifi.controller.UninheritableFlowException: Failed to connect node to cluster because local flow is different than cluster flow.
org.apache.nifi.controller.UninheritableFlowException: Failed to connect node to cluster because local flow is different than cluster flow.
	at org.apache.nifi.controller.StandardFlowService.loadFromConnectionResponse(StandardFlowService.java:921) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardFlowService.load(StandardFlowService.java:513) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.web.server.JettyServer.start(JettyServer.java:780) [nifi-jetty-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.NiFi.<init>(NiFi.java:162) [nifi-runtime-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.NiFi.main(NiFi.java:269) [nifi-runtime-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
Caused by: org.apache.nifi.controller.UninheritableFlowException: org.apache.nifi.processors.example.MyProcessor from org.apache.nifi:nifi-example-processors-nar:2.0 is not known to this NiFi instance.
	at org.apache.nifi.controller.StandardFlowSynchronizer.checkBundleCompatibility(StandardFlowSynchronizer.java:431) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardFlowSynchronizer.sync(StandardFlowSynchronizer.java:246) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.controller.FlowController.synchronize(FlowController.java:1535) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.persistence.StandardXMLFlowConfigurationDAO.load(StandardXMLFlowConfigurationDAO.java:83) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardFlowService.loadFromBytes(StandardFlowService.java:707) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardFlowService.loadFromConnectionResponse(StandardFlowService.java:896) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	... 4 common frames omitted
Caused by: java.lang.IllegalStateException: org.apache.nifi.processors.example.MyProcessor from org.apache.nifi:nifi-example-processors-nar:2.0 is not known to this NiFi instance.
	at org.apache.nifi.util.BundleUtils.findCompatibleBundle(BundleUtils.java:54) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.util.BundleUtils.getBundle(BundleUtils.java:88) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardFlowSynchronizer.checkBundleCompatibility(StandardFlowSynchronizer.java:429) ~[nifi-framework-core-1.2.0-SNAPSHOT.jar:1.2.0-SNAPSHOT]
	... 9 common frames omitted
```

I find this error message confusing. It used to make sense when the local flow as different than the cluster flow and that was the only way we could get into this situation. However, it now is caused by failing to create the necessary Processor. We should use a different error message.

Either we could throw a different Exception (perhaps something that extends UninheritableFlowException?) and provide a different error message, or we should make the message more generic, such as: "Unable to update local flow to match the cluster's flow" instead of "Local flow is different from cluster flow," given that the local flow was empty and therefore should try to inherit the cluster flow.

Does that make sense?

**Comment by bbende** [comment]
@markap14 The PR has been updated to address your comments and to fix a couple of other issues that were identified, also rebased against master as of yesterday evening.

Here are the changes since your last review:

- https://github.com/apache/nifi/pull/1585/commits/fadf43d7ccc755b5a5e3b52692e6b8f814afb657 (synchronization in node classes)
- https://github.com/apache/nifi/pull/1585/commits/b1ed76c2bc4ec3a639d360017fe899a8fd5777a7 (CS bundled with API + restructure Hive NAR)
- https://github.com/apache/nifi/pull/1585/commits/6b2ac1fb77ee08ce4b183e3b78c7c178c38256ae (missing components in fingerprinting)
- https://github.com/apache/nifi/pull/1585/commits/8988cd9a88ab528bbbb25c07260b17b5179c5886 (checkstyles and failing tests)

I'm going to continue testing today, but let me know if there is anything else to address.

**Comment by markap14** [comment]
This all looks good. I've run through the standard cases of standalone and clustered instances. Ran through some crazy cases that I could think of. Didn't run into any unexpected problems. Code looks good. I especially paid attention to the *Node classes, StandardFlowService, and Fingerprint classes when reviewing code but found nothing concerning. +1 merged to master!
